File: app/src/main/java/com/sdex/activityrunner/ApplicationsListAdapter.java
package com.sdex.activityrunner;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.recyclerview.extensions.ListAdapter;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.PopupMenu;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.RequestManager;
import com.bumptech.glide.request.RequestOptions;
import com.sdex.activityrunner.app.ActivitiesListActivity;
import com.sdex.activityrunner.db.application.ApplicationModel;
import com.sdex.activityrunner.glide.GlideApp;
import com.sdex.activityrunner.manifest.ManifestViewerActivity;
import com.sdex.activityrunner.util.IntentUtils;

public class ApplicationsListAdapter extends ListAdapter<ApplicationModel,
  ApplicationsListAdapter.AppViewHolder> {

  private final RequestManager glide;

  protected ApplicationsListAdapter(Context context) {
    super(DIFF_CALLBACK);
    this.glide = GlideApp.with(context);
  }

  @NonNull
  @Override
  public AppViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
    LayoutInflater inflater = LayoutInflater.from(parent.getContext());
    View view = inflater.inflate(R.layout.item_application, parent, false);
    return new AppViewHolder(view);
  }

  @Override
  public void onBindViewHolder(@NonNull AppViewHolder holder, int position) {
    holder.bindTo(getItem(position), glide);
  }

  static class AppViewHolder extends RecyclerView.ViewHolder {

    private final TextView name;
    private final TextView packageName;
    private final ImageView icon;
    private final View overflowMenu;

    public AppViewHolder(View itemView) {
      super(itemView);
      name = itemView.findViewById(android.R.id.text1);
      packageName = itemView.findViewById(android.R.id.text2);
      icon = itemView.findViewById(android.R.id.icon);
      overflowMenu = itemView.findViewById(R.id.app_menu);
    }

    public void bindTo(ApplicationModel item, RequestManager glide) {
      name.setText(item.getName());
      packageName.setText(item.getPackageName());

      Context context = itemView.getContext();

      glide.load(item)
        .apply(new RequestOptions().fitCenter())
        .into(icon);

      itemView.setOnClickListener(v -> ActivitiesListActivity.start(context, item));

      overflowMenu.setOnClickListener(v -> {
        PopupMenu popup = new PopupMenu(context, v);
        popup.inflate(R.menu.application_item_menu);
        popup.show();
        popup.setOnMenuItemClickListener(menuItem -> {
          final String packageName = item.getPackageName();
          switch (menuItem.getItemId()) {
            case R.id.action_open_app: {
              // TODO open the app
              return true;
            }
            case R.id.action_open_app_info: {
              IntentUtils.openApplicationInfo(context, packageName);
              return true;
            }
            case R.id.action_open_app_manifest: {
              ManifestViewerActivity.start(context, packageName, item.getName());
              return true;
            }
          }
          return false;
        });

      });
    }
  }

  public static final DiffUtil.ItemCallback<ApplicationModel> DIFF_CALLBACK =
    new DiffUtil.ItemCallback<ApplicationModel>() {

      @Override
      public boolean areItemsTheSame(ApplicationModel oldItem, ApplicationModel newItem) {
        return oldItem.getPackageName().equals(newItem.getPackageName());
      }

      @Override
      public boolean areContentsTheSame(ApplicationModel oldItem, ApplicationModel newItem) {
        return oldItem.equals(newItem);
      }
    };
}

File: app/src/main/java/com/sdex/activityrunner/app/ApplicationsListAdapter.kt
package com.sdex.activityrunner.app

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.core.view.isVisible
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.RequestManager
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions
import com.bumptech.glide.request.RequestOptions
import com.sdex.activityrunner.R
import com.sdex.activityrunner.databinding.ItemApplicationBinding
import com.sdex.activityrunner.db.cache.ApplicationModel
import com.sdex.activityrunner.preferences.AppPreferences
import com.sdex.activityrunner.util.ApplicationSectionNameProvider
import com.sdex.activityrunner.util.EmptySectionNameProvider
import com.sdex.activityrunner.util.SectionNameProvider
import com.simplecityapps.recyclerview_fastscroll.views.FastScrollRecyclerView

class ApplicationsListAdapter(
    activity: FragmentActivity,
    private val appPreferences: AppPreferences,
) : ListAdapter<ApplicationModel, ApplicationsListAdapter.AppViewHolder>(DIFF_CALLBACK),
    FastScrollRecyclerView.SectionedAdapter {

    private val glide = Glide.with(activity)

    private var showSystemAppIndicator: Boolean = appPreferences.isShowSystemAppIndicator
    private var showDisabledAppIndicator: Boolean = appPreferences.isShowDisabledAppIndicator
    private var sectionNameProvider: SectionNameProvider =
        if (appPreferences.sortBy == ApplicationModel.NAME) {
            ApplicationSectionNameProvider
        } else {
            EmptySectionNameProvider
        }
    var itemClickListener: ItemClickListener? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AppViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        return AppViewHolder(ItemApplicationBinding.inflate(inflater, parent, false))
    }

    override fun onBindViewHolder(holder: AppViewHolder, position: Int) {
        holder.bind(
            getItem(position),
            glide,
            showSystemAppIndicator,
            showDisabledAppIndicator,
            itemClickListener
        )
    }

    override fun getSectionName(position: Int): String =
        sectionNameProvider.getSectionName(getItem(position))

    @SuppressLint("NotifyDataSetChanged")
    fun update() {
        sectionNameProvider = if (appPreferences.sortBy == ApplicationModel.NAME) {
            ApplicationSectionNameProvider
        } else {
            EmptySectionNameProvider
        }
        if (appPreferences.isShowSystemAppIndicator != showSystemAppIndicator ||
            appPreferences.isShowDisabledAppIndicator != showDisabledAppIndicator
        ) {
            showSystemAppIndicator = appPreferences.isShowSystemAppIndicator
            showDisabledAppIndicator = appPreferences.isShowDisabledAppIndicator
            notifyDataSetChanged()
        }
    }

    interface ItemClickListener {

        fun onItemClick(item: ApplicationModel)

        fun onItemLongClick(item: ApplicationModel)
    }

    class AppViewHolder(
        private val binding: ItemApplicationBinding
    ) : RecyclerView.ViewHolder(binding.root) {

        fun bind(
            item: ApplicationModel,
            glide: RequestManager,
            showSystemAppIndicator: Boolean,
            showDisabledAppIndicator: Boolean,
            itemClickListener: ItemClickListener?,
        ) {
            binding.name.text = item.name
            binding.packageName.text = item.packageName
            val showSystemLabel = item.system && showSystemAppIndicator
            val showDisabledLabel = !item.enabled && showDisabledAppIndicator
            binding.info.isVisible = showDisabledLabel || showSystemLabel
            binding.system.isVisible = showSystemLabel
            binding.disabled.isVisible = showDisabledLabel

            val context = binding.root.context
            binding.version.text = context.getString(
                R.string.app_version_format,
                item.versionName,
                item.versionCode
            )

            val totalActivitiesFormattedText = context.resources.getQuantityString(
                R.plurals.activities_count,
                item.activitiesCount,
                item.activitiesCount,
            )
            binding.activities.text = context.getString(
                R.string.app_info_activities_number,
                totalActivitiesFormattedText,
                item.exportedActivitiesCount,
            )

            glide.load(item)
                .apply(RequestOptions().fitCenter())
                .transition(DrawableTransitionOptions.withCrossFade())
                .into(binding.icon)

            binding.root.setOnClickListener {
                itemClickListener?.onItemClick(item)
            }
            binding.root.setOnLongClickListener {
                itemClickListener?.onItemLongClick(item)
                true
            }
            binding.appMenu.setOnClickListener {
                itemClickListener?.onItemLongClick(item)
            }
        }
    }

    companion object {

        private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<ApplicationModel>() {

            override fun areItemsTheSame(
                oldItem: ApplicationModel,
                newItem: ApplicationModel
            ): Boolean {
                return oldItem.packageName == newItem.packageName
            }

            override fun areContentsTheSame(
                oldItem: ApplicationModel,
                newItem: ApplicationModel
            ): Boolean {
                return oldItem == newItem
            }
        }
    }
}
